tsdb/chunks/chunks.go

```diff
@@ -252,63 +254,144 @@ func MergeOverlappingChunks(chks []Meta) ([]Meta, error) {
 			last++
 			continue
 		}
-		nc := &newChks[last]
-		if c.MaxTime > nc.MaxTime {
-			nc.MaxTime = c.MaxTime
-		}
-		chk, err := MergeChunks(nc.Chunk, c.Chunk)
+		dividedChunkOne, dividedChunkTwo, err := MergeChunks(newChks[last].Chunk, c.Chunk, MaxSamplesInChunk)
 		if err != nil {
 			return nil, err
 		}
-		nc.Chunk = chk
+		mint, maxt := getMinAndMaxTimeOfChunk(dividedChunkOne)
+		newChks[last] = Meta{Chunk: dividedChunkOne, MinTime: mint, MaxTime: maxt}
+		if dividedChunkTwo != nil {
+			mint, maxt = getMinAndMaxTimeOfChunk(dividedChunkTwo)
+			newMeta := Meta{Chunk: dividedChunkTwo, MinTime: mint, MaxTime: maxt}
+			newChks = append(newChks, newMeta)
+			last++
+		}
 	}
-
 	return newChks, nil
 }
 
 // MergeChunks vertically merges a and b, i.e., if there is any sample
 // with same timestamp in both a and b, the sample in a is discarded.
-func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
+// Also breaks chunks that are larger than a particular limit.
+func MergeChunks(a, b chunkenc.Chunk, limit int) (chunkenc.Chunk, chunkenc.Chunk, error) {
+	newChunksArray := make([]chunkenc.Chunk, 0)
+	last := 0
 	newChunk := chunkenc.NewXORChunk()
 	app, err := newChunk.Appender()
+	appPtr := &app
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	ait := a.Iterator(nil)
 	bit := b.Iterator(nil)
 	aok, bok := ait.Next(), bit.Next()
+	numSamples := 0
+	appendAndReset := func(newChunksArrayPtr *[]chunkenc.Chunk, newChunk *chunkenc.XORChunk) (*chunkenc.XORChunk, *chunkenc.Appender, error) {
+		*newChunksArrayPtr = append(*newChunksArrayPtr, newChunk)
+		newChunk = chunkenc.NewXORChunk()
+		app, err = newChunk.Appender()
+		if err != nil {
+			return nil, nil, err
+		}
+		return newChunk, &app, err
+	}
 	for aok && bok {
 		at, av := ait.At()
 		bt, bv := bit.At()
 		if at < bt {
-			app.Append(at, av)
+			(*appPtr).Append(at, av)
+			numSamples++
+			if numSamples == limit {
+				newChunk, appPtr, err = appendAndReset(&newChunksArray, newChunk)
+				if err != nil {
+					return nil, nil, err
+				}
+				numSamples = 0
+			}
```

Comment:
Seems that you're doing the same thing in many places here, you could probably just move this section outside of the if statements and only do it once.

Reply:
Oops. Should have realized that earlier. My bad. Will fix it! Thanks for the review.